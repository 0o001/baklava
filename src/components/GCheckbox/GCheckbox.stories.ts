import { storiesOf } from '@storybook/vue';
import { withKnobs, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import GCheckbox from './';
import GBox from '../GBox';
import GText from '../GText';

storiesOf('GCheckbox', module)
  .addDecorator(withKnobs)
  .addParameters({ component: GCheckbox })
  .add('Single Select', () => ({
    components: { GCheckbox, GText },
    template: `
      <GCheckbox
        :value="{ id: 1 }"
        v-model="isChecked"
        @change="onChange"
        :disabled="disabled"
        :indeterminate="indeterminate"
      >
        <GText variant="body"> Is Selected: {{ isChecked }} </GText>
      </GCheckbox>
    `,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
      indeterminate: {
        default: boolean('Indeterminate', false),
      },
    },
    data () {
      return {
        isChecked: false,
      };
    },
    methods: {
      onChange (event) {
        action('Change')(event);
      },
    },
  }))
  .add('Multi Select with object values', () => ({
    components: { GCheckbox, GBox, GText },
    template: `
      <GBox>
        <GBox flex direction="row" :mb="10">
          <GCheckbox :value="value" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="{ id: 2 }" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="{ id: 3 }" v-model="selectedItems" :disabled="disabled"></GCheckbox>
        </GBox>
        <GText variant="body">Selected Items:</GText>
        <ul>
            <li v-for="selected in selectedItems"> {{ selected }} </li>
        </ul>
      </GBox>
    `,
    data () {
      const value = { id: 1 };
      return {
        value,
        selectedItems: [value],
      };
    },
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
  }))
  .add('Multi Select with string values', () => ({
    components: { GCheckbox, GBox, GText },
    template: `
      <GBox>
        <GBox flex direction="row" :mb="10">
          <GCheckbox :value="'value1'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="'value2'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
          <GCheckbox :value="'value3'" v-model="selectedItems" :disabled="disabled"></GCheckbox>
        </GBox>
        <GText variant="body">Selected Items:</GText>
        <ul>
            <li v-for="selected in selectedItems"> {{ selected }} </li>
        </ul>
      </GBox>
    `,
    data () {
      return {
        selectedItems: ['value2'],
      };
    },
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
    },
  }))
  .add('Props', () => ({
    components: { GCheckbox, GText },
    template: `
      <GCheckbox
        :value="{ id: 1 }"
        v-model="isChecked"
        @change="onChange"
        :disabled="disabled"
        :isInvalid="isInvalid"
        color="turquoise-500"
      >
        <GText variant="body"> Is Selected: {{ isChecked }} </GText>
      </GCheckbox>`,
    props: {
      disabled: {
        default: boolean('Disabled', false),
      },
      isInvalid: {
        default: boolean('Invalid', false),
      },
    },
    data () {
      return {
        isChecked: false,
      };
    },
    methods: {
      onChange (event) {
        action('Change')(event);
      },
    },
  }))
  .add('Events', () => ({
    components: { GCheckbox },
    template: '<GCheckbox :value="{ id: 1 }" @change="onChange" @click="onClick" />',
    methods: {
      onChange (event) {
        action('Changed')(event);
      },
      onClick (event) {
        action('Clicked')(event);
      },
    },
  }));
