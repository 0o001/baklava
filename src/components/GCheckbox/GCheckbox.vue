<template>
  <GBox
    flex
    direction="row"
    align="center"
  >
    <label class="g-checkbox">
      <input
        type="checkbox"
        class="-input"
        v-bind="$attrs"
        v-on="listeners"
        :value="value"
        :checked="isChecked"
        @change="toggleInput"
      >
      <span
        class="-inner"
        :class="{
          '-disabled': $attrs.disabled,
          '-invalid': isInvalid,
          '-checked': isChecked,
          [`g-bg-${color}`]:isChecked,
          [`g-bg-white`]:!isChecked
        }"
      >
        <GIcon
          size="16px"
          :name="$attrs.indeterminate ? 'minus' : 'check'"
          :stroke="strokeColor"
        />
      </span>
    </label>
    <GBox
      :ml="10"
      v-if="hasSlot"
    >
      <slot />
    </GBox>
  </GBox>
</template>

<script lang="ts">

import GBox from '../GBox/GBox.vue';
import GIcon from '../GIcon/GIcon.vue';

export default {
  name: 'ScCheckbox',
  model: {
    prop: 'modelValue',
    event: 'input',
  },
  components: {
    GBox,
    GIcon,
  },

  props: {
    value: {
      // type: Object | String,
      default: null,
    },
    modelValue: {
      // type: Array | Boolean,
      default: null,
    },
    isInvalid: {
      type: Boolean,
      default: false,
    },
    color: {
      type: String,
      default: 'orange-500',
    },
  },

  data () {
    return {
      isChecked: false,
    };
  },

  computed: {
    hasSlot () {
      return typeof this.$slots.default !== 'undefined';
    },
    listeners () {
      const { change, input, ...restListeners } = this.$listeners;

      return restListeners;
    },
    strokeColor () {
      if (this.$attrs.disabled && this.isChecked) {
        return 'var(--light-grey-500)';
      } else if (this.$attrs.disabled && !this.isChecked) {
        return 'var(--bg-grey-500)';
      } else {
        return 'var(--white)';
      }
    },
    hasSlot () {
      return typeof this.$slots.default !== 'undefined';
    },
  },

  methods: {
    toggleInput (event) {
      this.isChecked = event.target.checked;
      let value;

      if (Array.isArray(this.modelValue)) {
        value = [...this.modelValue];

        if (this.isChecked) {
          value.push(this.value);
        } else {
          value = value
            .filter(val => val.id
              ? val.id !== this.value.id
              : val !== this.value);
        }
      } else {
        value = this.isChecked;
      }

      this.$emit('change', { event, value: this.value, checked: this.isChecked });
      this.$emit('input', value);
    },
  },

  watch: {
    modelValue: {
      handler (newModelValue) {
        if (Array.isArray(newModelValue)) {
          if (newModelValue.indexOf(this.value) !== -1) {
            this.isChecked = true;
          } else if (newModelValue.length === 0) {
            this.isChecked = false;
          }
        } else {
          this.isChecked = newModelValue;
        }
      },
      immediate: true,
    },
  },
};
</script>

<style lang="scss" scoped>
  .g-checkbox {
    cursor: pointer;
    display: inline-block;
    position: relative;
    width: var(--spacing-16);
    height: var(--spacing-16);

    .-input {
      cursor: pointer;
      position: absolute;
      opacity: 0;
      top: var(--spacing-0);
      left: var(--spacing-0);
      width: var(--spacing-16);
      height: var(--spacing-16);
    }

    .-inner {
      text-align: center;
      width: var(--size-full);
      height: var(--size-full);
      border-radius: var(--radius-sm);
      display: flex;

      &.-checked {
        border: solid 1px transparent !important;
      }

      &.-disabled {
        background-color: var(--bg-grey-500);
        border: solid 1px var(--mid-grey-800) !important;
      }

      &.-invalid {
        border: solid 1px var(--red-500) !important;
      }

      &:not(.-invalid) {
        border: solid 1px var(--mid-grey-800);
      }
    }
  }
</style>
