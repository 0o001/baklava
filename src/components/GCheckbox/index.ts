import GCheckbox from './GCheckbox.vue';

export default GCheckbox;
