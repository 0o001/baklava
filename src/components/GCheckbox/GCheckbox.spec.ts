import { mount } from '@vue/test-utils';
import GCheckbox from '.';

describe('Checkbox', () => {
  let wrapper: any;
  let listeners: any;

  beforeEach(() => {
    listeners = {
      change: () => {},
      input: () => {},
      click: () => {},
    };
    wrapper = mount(GCheckbox, {
      propsData: {
        value: 'test',
      },
      listeners,
    });
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show icon-check if checkbox is checked', async () => {
    wrapper.find('.-input').trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.feather--check').exists()).toBe(true);
  });

  it('should have disabled class if disabled attribute is given', async () => {
    wrapper.setProps({
      disabled: true,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.-inner').classes('-disabled')).toBe(true);
  });

  it('should extract change and input event from $listeners', () => {
    const reducedListeners = wrapper.vm.listeners;

    expect(Object.keys(reducedListeners)).toEqual(['click']);
  });

  it('should emit change and input events on click checkbox', () => {
    wrapper.find('.-input').trigger('click');

    const { checked, event, value } = wrapper.emitted().change[0][0];
    const inputValue = wrapper.emitted().input[0][0];

    expect(checked).toEqual(true);
    expect(value).toEqual('test');
    expect(event.type).toEqual('change');
    expect(inputValue).toEqual(true);
  });

  it('should push given value to v-model array options if Array is given with v-model', async () => {
    wrapper.setProps({
      modelValue: [],
    });
    await wrapper.vm.$nextTick();

    wrapper.find('.-input').trigger('click');

    const { checked, event, value } = wrapper.emitted().change[0][0];
    const inputValue = wrapper.emitted().input[0][0];

    expect(checked).toEqual(true);
    expect(value).toEqual('test');
    expect(event.type).toEqual('change');
    expect(inputValue).toEqual(['test']);
  });

  it('should remove given value from v-model array options if Array is given with v-model', async () => {
    wrapper.setProps({
      modelValue: ['test', 'test2'],
    });
    await wrapper.vm.$nextTick();

    wrapper.vm.toggleInput({ target: { checked: false } });

    const inputValue = wrapper.emitted().input[0][0];

    expect(inputValue).toEqual(['test2']);
  });

  it('should remove given string value from v-model array options if Array is given with v-model', async () => {
    wrapper = mount(GCheckbox, {
      propsData: {
        value: 'test1',
        modelValue: ['test1', 'test2', 'test3'],
      },
      listeners,
    });

    wrapper.vm.toggleInput({ target: { checked: false } });

    wrapper.setProps({
      value: 'test2',
      modelValue: wrapper.emitted().input[0][0],
    });
    await wrapper.vm.$nextTick();

    wrapper.vm.toggleInput({ target: { checked: false } });

    wrapper.setProps({
      modelValue: wrapper.emitted().input[1][0],
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.modelValue).toEqual(['test3']);
  });

  it('should remove given object value from v-model array options if Array is given with v-model', async () => {
    wrapper = mount(GCheckbox, {
      propsData: {
        value: { id: 1 },
        modelValue: [{ id: 1 }, { id: 2 }, { id: 3 }],
      },
      listeners,
    });

    wrapper.vm.toggleInput({ target: { checked: false } });

    wrapper.setProps({
      value: { id: 2 },
      modelValue: wrapper.emitted().input[0][0],
    });
    await wrapper.vm.$nextTick();

    wrapper.vm.toggleInput({ target: { checked: false } });

    wrapper.setProps({
      modelValue: wrapper.emitted().input[1][0],
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.modelValue).toEqual([{ id: 3 }]);
  });

  it('should set true to isChecked if initial value is true', async () => {
    wrapper.setProps({ modelValue: true });

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isChecked).toBe(true);
  });

  it('should set true to isChecked if initial value is true', async () => {
    wrapper.setProps({ modelValue: true });

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isChecked).toBe(true);
  });

  it('should set true to isChecked if modelValue array contains a given value', async () => {
    const value = { id: 1 };

    wrapper = mount(GCheckbox, {
      propsData: {
        value,
        modelValue: [value],
      },
      listeners,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isChecked).toBe(true);
  });

  it('should set false to isChecked if modelValue array not contains a given value', async () => {
    const value = { id: 1 };

    wrapper = mount(GCheckbox, {
      propsData: {
        value,
        modelValue: [{ id: 2 }],
      },
      listeners,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isChecked).toBe(false);
  });

  it('should set false to isChecked if modelValue array is empty', async () => {
    const value = { id: 1 };

    wrapper = mount(GCheckbox, {
      propsData: {
        value,
        modelValue: [],
      },
      listeners,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.vm.isChecked).toBe(false);
  });

  it('should have is-invalid class if isInvalid prop is true', async () => {
    const value = { id: 1 };

    wrapper = mount(GCheckbox, {
      propsData: {
        value,
        modelValue: [],
        isInvalid: true,
      },
      listeners,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.-inner').classes('-invalid')).toBe(true);
  });

  it('should not have is-invalid class if isInvalid prop is false', async () => {
    const value = { id: 1 };

    wrapper = mount(GCheckbox, {
      propsData: {
        value,
        modelValue: [],
        isInvalid: false,
      },
      listeners,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.-inner').classes('-invalid')).not.toBe(true);
  });

  it('should set default color if do not given color as props', async () => {
    const defaultColor = '.g-bg-orange-500';

    wrapper.find('.-input').trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.find(defaultColor).exists()).toBe(true);
  });

  it('should set color if color given as props', async () => {
    const color = 'purple-500';

    wrapper.setProps({
      color: color,
    });

    wrapper.find('.-input').trigger('click');

    await wrapper.vm.$nextTick();

    expect(wrapper.find('.g-bg-purple-500').exists()).toBe(true);
  });
});
