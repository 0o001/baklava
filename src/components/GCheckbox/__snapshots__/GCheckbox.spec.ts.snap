// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<div
  class="g-d-flex g-d-align-center g-d-direction-row"
>
  <label
    class="g-checkbox"
  >
    <input
      class="-input"
      type="checkbox"
      value="test"
    />
     
    <span
      class="-inner"
    >
      <svg
        class="g-icon feather feather--check"
        dataName="check"
        dataTags=""
        dataType="check"
        fill="none"
        height="16px"
        stroke="var(--white)"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </span>
  </label>
   
  <!---->
</div>
`;
